### PR TITLE
Remove `logits.float()`

### PR DESCRIPTION
**Pull request by ringohoffman**
# What does this PR do?

Follow up to:

* https://github.com/huggingface/transformers/pull/31292
* https://github.com/huggingface/transformers/issues/30860

given that 4.45 has been released and 4.46 is next

Llama 3.1 8B FSDP2 peak inference memory usage with `float()` (18.5GiB):

float()

Llama 3.1 8B FSDP2 peak inference memory usage without `float()` (10.6GiB):

no_float()

In my environment, ~43% reduction in peak memory usage.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@Cyrilvallez @gante @ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
🚀🚀thanks for being so prompt in fixing! 🤗

**Comment by Cyrilvallez** [comment]
Hey @ringohoffman, I'm curious about the batch size you used to notice such a difference with the `float()` only, given that `num_logits_to_keep=1` by default?

**Comment by KyleMylonakisProtopia** [comment]
> batch size you used to notice such

I have observed these severe spikes with batch size 1. At this point I always just comment out that piece of code manually given it does not matter for my workflows at all.

**Comment by Cyrilvallez** [comment]
Humm, are you using `generate`? If not, are you making sure to pass `num_logits_to_keep=1` to the `forward`? By default it is 0 if not using `generate`. Because a batch size of 1 would result in a tensor of shape `(1, 1, 128k)` for Llama 3.1, which is always low independently of the dtype.

**Comment by KyleMylonakisProtopia** [comment]
We observe these spikes during training not generation, but we are using a different loss than the LM loss, so we don't require full precision here.

**Comment by Cyrilvallez** [comment]
I see, makes sense then, as long as you don't pass `labels` 🤗
